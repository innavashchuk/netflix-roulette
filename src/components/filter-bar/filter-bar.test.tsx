import * as React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import {Dispatch} from 'react';
import FilterBar from './filter-bar';

jest.mock('react-redux', () => ({
    useSelector: (cb: () => string) => cb(),
    useDispatch: () => (): null => null
}));
jest.mock('../../redux/selectors', () => ({
    ...jest.requireActual('../../redux/selectors'),
    selectFilterParam: (): string => 'Documentary',
    selectSortOrderParam: (): string => 'desc'
}));

const onFilterChange = jest.fn();

afterEach(() => {
    jest.clearAllMocks();
});

describe('FilterBar', () => {
    it('should initialize', () => {
        const {asFragment} = render(<FilterBar onFilterChange={onFilterChange}/>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should set correct filter value based on the useSelector(selectFilterParam)', () => {
        render(<FilterBar onFilterChange={onFilterChange}/>);
        const activeButton = screen.getByRole('button', {name: 'DOCUMENTARY'});
        expect(activeButton).toBeTruthy();
        expect(activeButton.classList).toContain('active');

        const inactiveButton = screen.getByRole('button', {
            name: 'ALL'
        });
        expect(inactiveButton.classList).not.toContain('active');
    });

    it('should set correct sort order value based on the useSelector(selectSortOrderParam)', () => {
        render(<FilterBar onFilterChange={onFilterChange}/>);
        const sortOrderIcon = screen.getByTestId('sortOrderIcon');
        expect(sortOrderIcon.className).toEqual('filter-tab__sort_direction_desc');
    });

    it('should call setState on button click', async () => {
        const state = 'desc';
        const setState: Dispatch<unknown> = jest.fn();
        const useStateSpy = jest.spyOn(React, 'useState')
        useStateSpy.mockImplementationOnce(() => [state, setState]);
        render(<FilterBar onFilterChange={onFilterChange}/>);
        const sortOrderButton = screen.getByRole('button', {name: 'SORT ORDER'});
        fireEvent.click(sortOrderButton);
        expect(setState).toHaveBeenCalledWith('asc');
    });
});
