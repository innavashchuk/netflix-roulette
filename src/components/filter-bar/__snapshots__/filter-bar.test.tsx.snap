// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar should initialize 1`] = `
<DocumentFragment>
  <div
    class="filter-tab"
  >
    <div
      class="filter-tab__filters"
    >
      <button
        class="filter-tab__item button-dark "
        id="filterKey0"
        name="All"
      >
        ALL
      </button>
      <button
        class="filter-tab__item button-dark active"
        id="filterKey1"
        name="Documentary"
      >
        DOCUMENTARY
      </button>
      <button
        class="filter-tab__item button-dark "
        id="filterKey2"
        name="Comedy"
      >
        COMEDY
      </button>
      <button
        class="filter-tab__item button-dark "
        id="filterKey3"
        name="Horror"
      >
        HORROR
      </button>
      <button
        class="filter-tab__item button-dark "
        id="filterKey4"
        name="Crime"
      >
        CRIME
      </button>
    </div>
    <div
      class="filter-tab__sort"
    >
      <p
        class="filter-tab__sort_lbl"
      >
        SORT BY
      </p>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline filter-tab__sort_btn button-dark"
      >
        <div
          aria-haspopup="listbox"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          role="button"
          tabindex="0"
        >
          Select...
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <button
        aria-label="SORT ORDER"
        class="filter-tab__sort_btn button-dark filter-tab__sort_direction"
      >
        <div
          class="filter-tab__sort_direction_desc"
          data-testid="sortOrderIcon"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
